genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/actions/SetUpJobAction.java

```diff
@@ -35,13 +59,354 @@
+            throw new SetUpJobException("Job directory parent is blank");
+        } else if (!parentDir.exists() || !parentDir.isDirectory()) {
+            throw new SetUpJobException("Invalid destination for job directory: " + parentDir);
+        } else if (!parentDir.isAbsolute()) {
+            throw new SetUpJobException("Job directory parent path is not absolute: " + parentDir);
+        }
+
+        final File jobDir = new File(parentDir, jobId);
+
+        if (jobDir.exists()) {
+            throw new SetUpJobException("Job directory already exists: " + jobDir);
+        }
+
+        try {
+            Files.createDirectory(jobDir.toPath());
+        } catch (final IOException e) {
+            throw new SetUpJobException("Failed to create job directory", e);
+        }
+
+        return jobDir;
+    }
+
+    private void createJobDirectoryStructure(
+        final JobSpecification jobSpec,
+        final File jobDirectory
+    ) throws SetUpJobException {
+
+        // Get DTOs
+        final List<JobSpecification.ExecutionResource> applications = jobSpec.getApplications();
+        final JobSpecification.ExecutionResource cluster = jobSpec.getCluster();
+        final JobSpecification.ExecutionResource command = jobSpec.getCommand();
+
+        // Make a list of all entity dirs
+        final List<Path> entityDirectories = Lists.newArrayList(
+            PathUtils.jobClusterDirectoryPath(jobDirectory, cluster.getId()),
+            PathUtils.jobCommandDirectoryPath(jobDirectory, command.getId()),
+            jobDirectory.toPath()
+        );
+
+        applications.stream()
+            .map(JobSpecification.ExecutionResource::getId)
+            .map(appId -> PathUtils.jobApplicationDirectoryPath(jobDirectory, appId))
+            .forEach(entityDirectories::add);
+
+        // Make a list of directories to create
+        // (only "leaf" paths, since createDirectories creates missing intermediate dirs
+        final List<Path> directoriesToCreate = Lists.newArrayList();
+
+        // Add config and dependencies for each entity
+        entityDirectories.forEach(
+            entityDirectory -> {
+                directoriesToCreate.add(PathUtils.jobEntityDependenciesPath(entityDirectory));
+                directoriesToCreate.add(PathUtils.jobEntityConfigPath(entityDirectory));
+            }
+        );
+
+        // Add logs dir
+        directoriesToCreate.add(PathUtils.jobGenieLogsDirectoryPath(jobDirectory));
+
+        // Create directories
+        for (final Path path : directoriesToCreate) {
+            try {
+                Files.createDirectories(path);
+            } catch (final Exception e) {
+                throw new SetUpJobException("Failed to create directory: " + path, e);
+            }
+        }
+    }
+
+    private DownloadManifest createDownloadManifest(
+        final File jobDirectory,
+        final JobSpecification jobSpec,
+        final List<File> setupFiles
```

Comment:
Why pass this in separately when it's already part of the job spec?

Reply:
`setupFiles` is an output parameter.
As I'm creating the manifest, I'm also noting down the setup files (since the latter are later sourced and used to build the process environment).

Reply:
I'm fine. Was just trying to save on passing around more parameters if they were already present. It's a private method so not a big deal if it ever needs to be refactored.